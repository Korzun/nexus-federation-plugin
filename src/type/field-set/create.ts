import { TypeCreate } from '../../builder';

import { model, Options as ModelOptions } from './model';

export type Options = ModelOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  builder.addOnInstallFunction((t) => {
    t.addType(model(options));
  });

  return builder;
};
