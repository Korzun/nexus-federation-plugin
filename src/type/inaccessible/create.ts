import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive } from './directive';
import { field } from './field';
import { inputObject } from './input-object';
import { object } from './object';

//TODO: Add missing type for argType, enum, enumValue
export const create: TypeCreate = (builder) => {
  builder.addOnInstallFunction((t) => {
    t.addType(directive);
  });

  /**
   * Adds `inaccessible` to Nexus field definitions and generates the
   * `@inaccessible` directive when it is detected in a field definition
   */
  builder.addFieldDefType(field);
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.inaccessible) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  /**
   * Adds `inaccessible` to Nexus object definitions and generates the
   * `@inaccessible` directive when it is detected in an object definition
   */
  builder.addObjectDefType(object);
  builder.addOnObjectDefinitionFunction((t, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.inaccessible) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  /**
   * Adds `inaccessible` to Nexus input field definitions and generates the
   * `@inaccessible` directive when it is detected in an input field definition.
   *
   * For some reason it is not necessary to call `addInputFieldDefType` to add
   * the input field. The type added through `addFieldDefType` appears to be sufficient
   */
  builder.addOnAddInputFieldFunction((inputFieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (inputFieldConfig.inaccessible) {
      if (!inputFieldConfig.directives) {
        inputFieldConfig.directives = new Array();
      }
      (inputFieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  //FIXME: Nexus Bug - the key `inaccessible` isn't properly surfaced in onAdd function
  // /**
  //  * Adds `inaccessible` to Nexus input object definitions and generates the
  //  * `@inaccessible` directive when it is detected in an input object definition
  //  */
  builder.addInputObjectDefType(inputObject);
  // builder.addOnInputObjectDefinitionFunction((inputObjectConfig) => {
  //   if (inputObjectConfig.inaccessible) {
  //     if (!inputObjectConfig.directives) {
  //       inputObjectConfig.directives = new Array();
  //     }
  //     (inputObjectConfig.directives as Array<core.NexusDirectiveUse>).push(
  //       addDirective('inaccessible', {}),
  //     );
  //   }
  // });

  return builder;
};
