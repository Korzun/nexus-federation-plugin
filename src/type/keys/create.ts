import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { object, Options as ObjectOptions } from './object';

export type Options = DirectiveOptions & ObjectOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@key` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `keys` to Nexus object definitions and generates the `@key` directive
   * when it is detected in a object definition
   */
  builder.addObjectDefType(object(options));
  builder.addOnObjectDefinitionFunction((block, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.keys) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('key', { fields: objectConfig.keys.join(' ') }),
      );
    }
  });

  return builder;
};
