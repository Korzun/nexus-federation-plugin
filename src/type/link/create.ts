import { TypeCreate } from '../../builder';

import { directive } from './directive';

export const create: TypeCreate = (builder) => {
  builder.addOnInstallFunction((t) => {
    t.addType(directive);
  });

  return builder;
};
