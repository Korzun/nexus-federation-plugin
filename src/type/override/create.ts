import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';

export type Options = DirectiveOptions & FieldOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@override` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `override` to Nexus field definitions and generates the `@override`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.override) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('override', { from: fieldConfig.override }),
      );
    }
  });

  return builder;
};
