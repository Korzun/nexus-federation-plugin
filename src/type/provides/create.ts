import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';

export type Options = DirectiveOptions & FieldOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@provides` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `provides` to Nexus field definitions and generates the `@provides`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.provides) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('provides', { fields: fieldConfig.provides.join(' ') }),
      );
    }
  });

  return builder;
};
