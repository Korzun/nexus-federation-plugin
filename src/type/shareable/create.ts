import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';
import { object, Options as ObjectOptions } from './object';

export type Options = DirectiveOptions & FieldOptions & ObjectOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@shareable` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `shareable` to Nexus object definitions and generates the `@shareable`
   * directive when it is detected in an object definition
   */
  builder.addObjectDefType(object(options));
  builder.addOnObjectDefinitionFunction((t, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.shareable) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('shareable', {}),
      );
    }
  });

  /**
   * Adds `shareable` to Nexus field definitions and generates the `@shareable`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.shareable) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('shareable', {}),
      );
    }
  });

  return builder;
};
