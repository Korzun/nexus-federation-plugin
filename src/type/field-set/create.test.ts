import * as testHelper from '../../test';

import { create } from './create';

describe('field set', () => {
  describe('model', () => {
    it('is added to schema', () => {
      const schema = testHelper.createSchema(create);
      expect(schema).toMatchSnapshot();
    });

    it('is added to types', async () => {
      const typegen = await testHelper.createTypegen(create);
      expect(typegen).toMatchSnapshot();
    });
  });
});
