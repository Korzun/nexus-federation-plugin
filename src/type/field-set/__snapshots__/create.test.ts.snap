// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field set model is added to schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


type Query {
  ok: Boolean!
}

\\"\\"\\"

  Represents a set of fields. Grammatically, a field set
  is a selection set minus the braces. This means it can
  represent a single field \\"upc\\", multiple fields \\"id
  countryCode\\", and even nested selection sets \\"id organization
  { id }\\".
  
  [FieldSet - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/federation-spec/#scalar-fieldset)
  
\\"\\"\\"
scalar _FieldSet"
`;

exports[`field set model is added to types 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */







declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  _FieldSet: any
}

export interface NexusGenObjects {
  Query: {};
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars

export interface NexusGenFieldTypes {
  Query: { // field return type
    ok: boolean; // Boolean!
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    ok: 'Boolean'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenDirectives = never

export interface NexusGenDirectiveArgs {
}

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    resolveType: true
    __typename: true
    isTypeOf: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  directives: NexusGenDirectives;
  directiveArgs: NexusGenDirectiveArgs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
    
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
    
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
    
  }
}"
`;
